torchdata/dataloader2/dataloader2.py

```diff
@@ -377,25 +377,31 @@ def _restore_checkpoint_beginning_of_epoch(self) -> None:
         """
         self._seed_generator = self._initial_seed_generator
 
-    def _pause(self):
+    def _pause(self) -> None:
         if hasattr(self.reading_service, "_pause"):
             self._is_paused = True
-            self.reading_service._pause()
+            pause_fn = self.reading_service._pause()
+            if pause_fn is not None:
+                self.datapipe = pause_fn(self.datapipe)
         else:
             warnings.warn("ReadingService doesn't support `pause`.")
 
-    def _resume(self):
+    def _resume(self) -> None:
         if hasattr(self.reading_service, "_resume"):
             if not self._is_paused:
                 warnings.warn("Resume is called when `DataLoader2` is not paused. No operation is performed.")
             else:
-                self.reading_service._resume()
+                resume_fn = self.reading_service._resume()
+                if resume_fn is not None:
+                    self.datapipe = resume_fn(self.datapipe)
                 self._is_paused = False
         else:
             warnings.warn("ReadingService doesn't support `resume`.")
 
     def _limit(self, num_batches: Optional[int]) -> None:
         if hasattr(self.reading_service, "_limit"):
-            self.reading_service._limit(num_batches)
+            limit_fn = self.reading_service._limit(num_batches)
+            if limit_fn is not None:
+                self.datapipe = limit_fn(self.datapipe, num_batches)
```

Comment:
Depending on the `limit_fn`, the `limit` on the DataPipe and `DataLoader2Iterator` can be out of sync. One example is if there exists a `filter` in the pipeline. Correct?

Reply:
Correct. I will add a note in `limit` that miniepoch doesn't support `filter` operation when dispatching process gets involved. It is still fine without dispatching process (only workers).